cpp/include/cudf/strings/detail/pad_impl.cuh

```diff
@@ -0,0 +1,126 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <cudf/strings/detail/utf8.hpp>
+#include <cudf/strings/detail/utilities.cuh>
+#include <cudf/strings/side_type.hpp>
+#include <cudf/strings/string_view.cuh>
+
+namespace cudf {
+namespace strings {
+namespace detail {
+
+/**
+ * @brief Return the size in bytes of padding d_str to width characters using a fill character
+ * with byte length of fill_char_size
+ *
+ * Pad does not perform truncation. That is, If `d_str.length() > width` then `d_str.size_bytes()`
+ * is returned.
+ *
+ * @param d_str String to pad
+ * @param width Number of characters for the padded string result
+ * @param fill_char_size Size of the fill character in bytes
+ * @return The number of bytes required for the pad
+ */
+__device__ size_type compute_padded_size(string_view d_str,
+                                         size_type width,
+                                         size_type fill_char_size)
+{
+  auto const length = d_str.length();
+  auto bytes        = d_str.size_bytes();
+  if (width > length)                            // no truncating;
+    bytes += fill_char_size * (width - length);  // add padding
+  return bytes;
+}
+
+/**
+ * @brief Pad d_str with fill_char into output up to width characters
+ *
+ * Pad does not perform truncation. That is, If `d_str.length() > width` then
+ * then d_str is copied into output.
+ *
+ * @tparam side Specifies where fill_char is added to d_str
+ * @param d_str String to pad
+ * @param width Number of characters for the padded string result
+ * @param fill_char Size of the fill character in bytes
+ * @param output Device memory to copy the padded string into
+ */
+template <side_type side = side_type::RIGHT>
+__device__ void pad_impl(cudf::string_view d_str,
+                         cudf::size_type width,
+                         cudf::char_utf8 fill_char,
+                         char* output)
+{
+  auto length = d_str.length();
+  if constexpr (side == side_type::LEFT) {
+    while (length++ < width) {
+      output += from_char_utf8(fill_char, output);
+    }
+    copy_string(output, d_str);
+  }
+  if constexpr (side == side_type::RIGHT) {
+    output = copy_string(output, d_str);
+    while (length++ < width) {
+      output += from_char_utf8(fill_char, output);
+    }
+  }
+  if constexpr (side == side_type::BOTH) {
+    auto const pad_size = width - length;
+    // an odd width will right-justify
+    auto right_pad = (width & 1) ? pad_size / 2 : (pad_size - pad_size / 2);
```

Comment:
nit: I personally prefer regular arithmetic over bitwise operations for even/odd tests, but that may be more verbose
```suggestion
    auto right_pad = (width % 2 != 0) ? pad_size / 2 : (pad_size - pad_size / 2);
```
alternatively: `pad_size / 2 + 1 - (width % 2)`

Reply:
The original idea seems to be overcomplicated(?) because: if not `width & 1` (even) then `pad_size - pad_size / 2 == pad_size / 2`. So why not just `auto right_pad = pad_size / 2`?

Reply:
`width` and `length` can be even/odd independently, so the same is true for `pad_size` and `width`

Reply:
> `width` and `length` can be even/odd independently, so the same is true for `pad_size` and `width`

Oh I see.

Reply:
I guess I do not agree that the proposed solution is clearer. I had to write it out to see if it was actually correct. I suspect Nghia did too based on his comments here.
The original code seems simpler to understand: if width is odd, adjust the `right_pad` (just like the comment says).
Also, the original implementation appears to be 2 instructions less than proposed solution: https://godbolt.org/z/37dzfGe9j

Reply:
It took a bit for me to understand that both the parity of `width` and `pad_width` influence the rounding here. My suggestion would be to make the dependency on both parities clearer, but I understand that the above code may not reach that goal. replacing `& 1` by `% 2` already makes things a bit clearer IMO.